hs/src/Reach/Eval/Core.hs

```diff
@@ -3915,7 +3921,61 @@ evalStmt = \case
           SLV_DLVar {} -> select_all de_val
           _ -> impossible "switch mvar"
       locAtf (srcloc_after_semi "switch" a sp) $ retSeqn fr ks_ne
-  (s@(JSThrow a _ _) : _) -> illegal a s "throw"
+  ((JSThrow a e _) : _) -> do
+    (dv_ty, dv) <- locAtf (srcloc_jsa "throw" a) (evalExpr e) >>= compileTypeOf . snd
+    curSt <- readSt id
+    exn_ref <- asks e_exn
+    exn <- liftIO $ readIORef exn_ref
+    -- Ensure we are throwing inside of a `try` block
+    unless (e_exn_in_throw exn) $ expect_ Err_Throw_No_Catch
+    -- Set or ensure the type that this `try` block is expected to catch
+    case e_exn_ty exn of
+      Nothing -> liftIO $ modifyIORef exn_ref (\ ex -> ex { e_exn_ty = Just dv_ty })
+      Just ty
+        | ty /= dv_ty -> expect_ $ Err_Try_Type_Mismatch ty dv_ty
+        | otherwise   -> return ()
+    saveLift =<< withAt (`DLS_Throw` dv)
+    liftIO $ modifyIORef exn_ref (\ ex -> ex { e_exn_st = Just curSt })
+    st <- asks e_st
+    liftIO $ modifyIORef st (\ s -> s { st_live = False })
+    asks (SLStmtRes . e_sco) <*> pure []
+  ((JSTry try_a (JSBlock _ stmts _) [JSCatch _ _ ce _ (JSBlock _ handler _)] _) : ks) -> do
+    locAtf (srcloc_jsa "try" try_a) $ do
+      at <- withAt id
+      -- Create fresh exception environment
+      exn_env_ref <-
+        liftIO $ newIORef $ ExnEnv
+          { e_exn_in_throw = True
+          , e_exn_ty = Nothing
+          , e_exn_st = Nothing}
+      let locTry = local (\e -> e { e_exn = exn_env_ref })
+      -- Process try block
+      SLRes try_stmts try_st try_ret <- captureRes $ locTry $ evalStmt stmts
+      exn_env <- liftIO $ readIORef exn_env_ref
+      case exn_env of
+        -- Get the type of thrown/caught expression & the state during throw
+        ExnEnv { e_exn_ty = Just arg_ty, e_exn_st = Just thrown_st } -> do
+            -- Bind `catch` argument before evaluating handler
+            handler_arg <- ctxt_mkvar (DLVar at Nothing arg_ty)
+            handler_env <- evalDeclLHS True Public mempty (SLV_DLVar handler_arg) ce
+            sco <- asks e_sco
+            let sco_env' = handler_env `M.union` sco_cenv sco
```

Comment:
This should use `envMerge` which protects against double binding, but really the `mempty` on line 3960 should be `sco_cenv sco`